Move getGlobalPlacement() and fix alignment direction for transformed objects

Rexbas opened the pull request:
Partially fixes #14576.

The alignment direction was not correct for objects with a transformation. For example, making a simple box in Part Design and then using the transform tool on the body to rotate the box resulted in wrong alignment directions. This is fixed by considering the orientation of the object. With this fix, the files provided in #14576 now align the view with the selected faces. 

However, it does not work properly yet when the assembly contains linked objects, like in AssemblyExample.FCStd. After using the transform tool to rotate the assembly, selecting a face and trying to align does not align with the selected face. The view is instead aligned with the face of the linked object which has another orientation. I don´t know why that happens but I will look into it another time.

wwmayer commented:
> However, it does not work properly yet when the assembly contains linked objects, like in AssemblyExample.FCStd.

This cannot be done inside the Part::Feature class because it doesn't know the link object it is referenced by. The right function to handle link objects is inside `View3DInventorViewer::alignToSelection()` because there you get the link object directly from the selection.

AFAICS there is currently no function to get the global placement of a link object. So, what's needed is something similar to `GeoFeature::globalPlacement()` .

In order to get things working (it worked for me in the above scenarios) revert your changes made to `Feature::getCameraAlignmentDirection` and try this instead:
```
#include <App/GeoFeatureGroupExtension.h>
void View3DInventorViewer::alignToSelection()
{
    if (!getCamera()) {
        return;
    }

    const auto selection = Selection().getSelection();

    // Empty selection
    if (selection.empty()) {
        return;
    }

    // Too much selections
    if (selection.size() > 1) {
        return;
    }

    // Get the geo feature
    App::GeoFeature* geoFeature = nullptr;
    std::pair<std::string, std::string> elementName;
    App::GeoFeature::resolveElement(selection[0].pObject, selection[0].SubName, elementName, false, App::GeoFeature::ElementNameType::Normal, nullptr, nullptr, &geoFeature);
    if (!geoFeature) {
        return;
    }

    Base::Placement grpPlm;
    if (auto* group = App::GeoFeatureGroupExtension::getGroupOfObject(selection[0].pObject)) {
        auto ext = group->getExtensionByType<App::GeoFeatureGroupExtension>();
        grpPlm = ext->globalGroupPlacement();
    }
    if (auto link = dynamic_cast<App::Link*>(selection[0].pObject)) {
        Base::Matrix4D mat = link->getTransform(true);
        Base::Placement plm(mat);
        grpPlm = grpPlm * plm;
    }

    Base::Rotation grpRot = grpPlm.getRotation();

    Base::Vector3d direction;
    if (geoFeature->getCameraAlignmentDirection(direction, selection[0].SubName)) {
        grpRot.multVec(direction, direction);
        const auto orientation = SbRotation(SbVec3f(0, 0, 1), Base::convertTo<SbVec3f>(direction));
        setCameraOrientation(orientation);
    }
}
```
Before updating your PR you have to refactor this code:
1. Add the virtual method
```
Base::Placement DocumentObject::globalGroupPlacement() const
{
    Base::Placement grpPlm;
    if (auto* group = App::GeoFeatureGroupExtension::getGroupOfObject(selection[0].pObject)) {
        auto ext = group->getExtensionByType<App::GeoFeatureGroupExtension>();
        grpPlm = ext->globalGroupPlacement();
    }
    return grpPlm;
}
```
2. Override the method in App::Link:
```
Base::Placement Link::globalGroupPlacement() const
{
    Base::Placement grpPlm = DocumentObject::globalGroupPlacement();
    Base::Placement plm(getTransform(true));
    return grpPlm * plm;
}
```

Then you can write in `View3DInventorViewer::alignToSelection()`
```
...
auto object = selection[0].pObject;
Base::Placement grpPlm = object->globalGroupPlacement();
...
```

Rexbas commented:
Thanks @wwmayer, I tried your suggestion and it works when the assembly is transformed but not when the part the link references has a different non-zero orientation. The alignment direction is still determined on the original object and not on the link. So we have to multiply with the inverse placement of the original object to get the correct alignment direction. With these changes, AssemblyExample.FCStd works well with linked objects.

I moved `GeoFeature::globalPlacement()` to `DocumentObject::globalPlacement()` which GeoFeature now overrides. The behavior should be the same. Overriding in GeoFeature is probably not necessary but I kept it for now to make sure I don't break things.

There is still one situation known to me where the alignment direction is wrong. I attached LinkAssemblyAlignment.FCStd which contains a 45 degree rotated body. This body is then linked to in an assembly. I noticed that this code, inserted in `View3DInventorViewer::alignToSelection()`: 
```
if (auto link = dynamic_cast<App::Link*>(selection[0].pObject)) {
    std::cout << "Hello link" << std::endl;
}
```
only runs in AssemblyExample.FCStd but not in LinkAssemblyAlignment.FCStd. I am still investigating why it works in one and not the other.

[LinkAssemblyAlignment.FCStd.zip](https://github.com/user-attachments/files/16238534/LinkAssemblyAlignment.FCStd.zip)

wwmayer commented on src/App/DocumentObject.cpp:
```diff
@@ -1416,6 +1416,19 @@ bool DocumentObject::redirectSubName(std::ostringstream &, DocumentObject *, Doc
     return false;
 }
 
+Base::Placement DocumentObject::globalPlacement() const {
```
This is a code smell because a base class should not make assumptions about a sub-class.

Rexbas replied:
I assume its because I use the Placement in that method? I changed the PR and it now includes the `globalGroupPlacement()` you suggested in the comment from a few days ago, assuming that is not a code smell.

wwmayer replied:
For some explanations read this: https://luzkan.github.io/smells/base-class-depends-on-subclass
A further problem would be that whenever using `globalPlacement()` it will never be clear what the returned placement will be and thus is suspicious to also break the Liskov substitution principle.

Rexbas replied:
Thanks, nice link. I guess this is resolved then.

Rexbas commented:
I found one more issue when the link transform property is true. This is also fixed.

I think the alignment direction for the assembly I attached in my previous comment is because the selection is not an `App::Link` but a `PartDesign::Pad`. It points to the same PartDesign::Pad as the original pad and thus uses the placement and group placement of the original pad. It seems a sub feature of a link is not a link to the sub feature of the original object?! Maybe an issue in the link, or selection? Seems like a different problem so I think this PR is ready.

sliptonic commented:
@PaddleStroke could you take a quick look and comment?  I'm wondering if there's interaction with other work being done related to global placement

PaddleStroke commented:
Yes I think this align to feature requires a proper getGlobalPlacement function. Because you need the global placement to find the correct direction.
Assembly workbench now have a getGlobalPlacement that should work properly now. It needs the full path to the selected object, ie the object + sub that is given by the selection observer.

PaddleStroke commented:
@Rexbas you can check the new assembly getGlobalPlacement, it merged yesterday. It's in assemblyobject.cpp as static functions.
Perhaps this should be moved to a Core location.

wwmayer commented:
> I found one more issue when the link transform property is true. This is also fixed.

I can't see a different behaviout between your new implementation and my suggestion. Or do I miss something?

Currently both implementations fail if a linked body has an identity placement while the referenced body has a non-identity placement.

wwmayer commented:
> you can check the new assembly getGlobalPlacement

What should be the arguments for targetObj, rootObj and sub if the user has selected a sub-shape of an object?

PaddleStroke commented:
RootObj and sub are what the selection observer give with no resolve. (Without the tnp string)
Targetobj is the object of which we want the placement.

So for example :
Assembly
- part
- - body
- - - pad
- - - - sketch

If you select the pad selection observer gives
RootObject = assembly
Sub = "part.body.pad.edge1"

Then target object is the object of which you want the placement. So either part or body for example.

Rexbas commented:
> Currently both implementations fail if a linked body has an identity placement while the referenced body has a non-identity placement.

Hmm I see it too now... . Lets see if the new getGlobalPlacement works.

wwmayer commented:
> RootObj and sub are what the selection observer give with no resolve. (Without the tnp string)

Thanks but it doesn't really help me to get further.

When I implement this simplified selection observer
```
class MyObs:
  def addSelection(self, doc, obj, sub, pnt):
    print ("Object = {}\nElement = {}".format(obj, sub))

obs = MyObs()
Gui.Selection.addObserver(obs, ResolveMode.NoResolve)
```
and use it for LinkAssemblyAlignment.FCStd then I get this output when I click on the pad's edge
```
Object = Body
Element = Pad.;g374;SKT;:Had0,E;:G;XTR;:Had0:7,F;:U;XTR;:Had0:7,E.Edge12
```
and this when I click on the same edge of the linked object
```
Object = Assembly
Body001.Pad.;g374;SKT;:Had0,E;:G;XTR;:Had0:7,F;:U;XTR;:Had0:7,E.Edge12
```
So, the root object will be Body or Assembly, respectively. But what would be the target object? Or is it equal to the root object or even null? And how do I get rid of the TNP naming?

PaddleStroke commented:
The target object can be the same as the root object.
If target obj is null I think (I'm currently away) that it takes the latest feature.
I have code in assembly that removes the tnp string. Though I can't say on top of my head where this is. Probably in the jointObject.py task' addSelection.

wwmayer commented:
The method in Python to resolve TNP subnames is resolveSubElement (in C++ it's GeoFeature::resolveElement using the oldName of the passed ElementNamePair).

wwmayer commented:
A much easier way is to not try to get the global placement from the document but let the viewer do it because it knows of all the SoTransform nodes. The only downside is that this way it won't support edges any more.
```
void View3DInventorViewer::alignToSelection()
{
    if (!getCamera()) {
        return;
    }

    const auto selection = Selection().getSelection();

    // Wrong selection
    if (selection.size() != 1) {
        return;
    }

    // Get the geo feature
    App::GeoFeature* geoFeature = nullptr;
    App::ElementNamePair elementName;
    App::GeoFeature::resolveElement(selection[0].pObject, selection[0].SubName, elementName, false, App::GeoFeature::ElementNameType::Normal, nullptr, nullptr, &geoFeature);
    if (!geoFeature) {
        return;
    }

    Base::Vector3d direction;
    if (!geoFeature->getCameraAlignmentDirection(direction, selection[0].SubName)) {
        return;
    }

    const SbVec3f pos(selection[0].x, selection[0].y, selection[0].z);
    const SbVec3f dir = getViewDirection();

    SoRayPickAction rp(getSoRenderManager()->getViewportRegion());
    rp.setRay(pos, dir);
    rp.setRadius(getPickRadius());
    rp.apply(getSceneGraph());

    if (SoPickedPoint* pick = rp.getPickedPoint()) {
        if (const SoDetail* detail = pick->getDetail()) {
            if (detail->isOfType(SoFaceDetail::getClassTypeId())) {
                SbVec3f normal = pick->getNormal();
                const auto orientation = SbRotation(SbVec3f(0, 0, 1), normal);
                setCameraOrientation(orientation);
            }
        }
    }
}
```

Rexbas commented:
@PaddleStroke, If either `targetObj` or `rootObj` is null then the `getGlobalPlacement` returns `Base::Placement()`. So I am passing the the same object to both the `targetObj` and the `rootObj`. The problem with this is that the rotation angle in the returned placement is twice as big as the actual angle (because root placement is multiplied with target placement which are the same in this case). If I scale the angle by 0.5 after calling the `getGlobalPlacement` (and some other modifications in `View3DInventorViewer::alignToSelection`) then it actually works really well, and it also still works for edges. So I would like to use this method but it needs some minor modifications. Also the method should be moved to a more central location. Would you like to do this?

PaddleStroke commented:
@wwmayer ah I didn't found the cpp implementation of the resolveElement. I will look at GeoFeature::resolveElement

@Rexbas Yes this needs to be at a more central location indeed. But I am not sure where this would fit best. Regarding the issue that you have, I think this is because in assembly I always have at least Assembly.Cube, so the rootObj is never the same as the targetObj. So this case may not be working properly indeed.

Rexbas commented:
> suggestion where we could implement such getGlobalPlacement?

What about App/Placement.cpp? That file is currently almost empty, I think it makes sense to have the static global placement method included there as well. And then we can use `Placement::getGlobalPlacement()`.

Rexbas commented:
I have moved `splitSubName()` from `AssemblyObject` to `DocumentObject` and `getGlobalPlacement()` from `AssemblyObject` to `App::Placement`. I also changed `getGlobalPlacement()` to return the root placement when root object and target object are the same. Alignment seems to work well now.

yorikvanhavre commented:
@wwmayer and @PaddleStroke could you please have a new look if this seems ok for you?

PaddleStroke commented:
For me this looks good. getGlobalPlacement definitely belong to core. I just wasn't sure where to put it. This place seems ok to me but this is for @wwmayer to confirm.

wwmayer commented on src/App/DocumentObject.cpp:
```diff
@@ -1412,6 +1412,27 @@ const char *DocumentObject::hasHiddenMarker(const char *subname) {
     return hiddenMarker()==marker?marker:nullptr;
 }
 
+std::vector<std::string> DocumentObject::splitSubName(const std::string& subname)
```
I would rather move this function to the Tools helper class in FreeCADBase.

wwmayer commented on src/App/Placement.cpp:
```diff
@@ -41,7 +43,65 @@ Placement::Placement() = default;
 
 Placement::~Placement() = default;
 
+Base::Placement Placement::getPlacementFromProp(App::DocumentObject* obj, const char* propName)
```
The purpose of App::Placement doesn't really fit to these helper functions. I would rather move them to GeoFeature or a new helper class.

Rexbas replied:
Moved to GeoFeature

yorikvanhavre commented:
@wwmayer is it good to go for you?

Rexbas commented:
There is a conflict because `getGlobalPlacement()` was modified. Now I also need to move the static `isLink(App::DocumentObject* obj)` away from `AssemblyObject`. I am thinking to make it a non-static method of `App::DocumentObject`. What do you think @wwmayer, good idea or do you have another suggestion?

PaddleStroke commented:
The goal of the isLink function that I added was to tell appart link groups from normal links. Because both are App::Link. On top of that LinkElements also need to return true.

Rexbas commented:
@PaddleStroke, do you have a simple test document containing a normal link, a group link and a LinkElement? I can't find/create a situation where `isLinkGroup()` returns true.

Rexbas commented:
I have moved `isLink()` and `isLinkGroup()` to `DocumentObject`. @PaddleStroke, please review and confirm that the returned values of these methods are as expected.

PaddleStroke commented:
[linkgroup.zip](https://github.com/user-attachments/files/16927249/linkgroup.zip)
Here's a sample.
It looks good to me. But I think it's for @wwmayer to confirm if it's ok.

wwmayer reviewed (commented):
It's OK now

Rexbas commented:
@chennes, this PR has 5 commits but I see only 4 in the 1.0 branch. Did you forget to cherry pick 8db7191084352cc83e908b3819cec3612058ccd2?

PaddleStroke commented:
@wwmayer the new getGlobalPlacement has no python binding. I am not sure how to add it since : 
- geoFeature already has a getGlobalPlacement in python, but it is the old globalPlacement function.
- It's a static function. ie it should be called without an instance of object. So something like App.getGlobalPlacement(...)

Any idea? Thanks

wwmayer commented:
Add this to GeoFeaturePy.xml: `<Methode Name="getGlobalPlacementOf" Static="true">` Then in Python you can call it as `App.GeoFeature.getGlobalPlacementOf`.

PaddleStroke commented:
Thanks, will do.
